File: Server/src/model/Spell/HealSpell.java
package model.Spell;
import model.Battle.BattleObject;
import model.Define.ModelDefine;
import model.Monster.Monster;
import model.Tower.Tower;
import java.util.Hashtable;
import java.util.Iterator;
import java.util.LinkedList;

public class HealSpell extends Spell{

    private Hashtable<Monster, Double> monsters;
    private LinkedList<Monster> targets;
    public HealSpell(int level) {
        super(2, 1);
        readSpellJson();
        targets = new LinkedList<>();
    }
    protected void readSpellJson(){
        super.readSpellJson();
        this.leftTick = (int)this.duration * ModelDefine.TICK_PER_SECOND;
        this.effectTick = 4 * ModelDefine.TICK_PER_SECOND;
    }
    public boolean update(LinkedList<Monster>monsters, LinkedList<Tower>towers){
        System.out.println("Spell " + battleObjectId + " time left: " + leftTick);
        if(leftTick > 0){
            Iterator<Monster> itr = monsters.iterator();
            while(itr.hasNext()) {
                Monster monster = itr.next();
                if (curPos.getDistance(monster.getCurPos()) <= range && monster.isBuffBySpell(this) == false) {
                    monster.addSpell(this);
                }
            }
            leftTick -= 1;
            return true;
        }
        else {
            return false;
        }
    }
    public void doAction(BattleObject battleObject){
        Monster monster = (Monster) battleObject;
        if(monster.getEffectTick(this) % 6 == 0){
            monster.heal(2);
        }
    }

    public void removeEffect(BattleObject battleObject){

    }
}

File: Client/src/modules/models/spell/HealSpell.js
var HealSpell = Spell.extend({
    ctor:function(){
        this._super(2);
        this.effectTick = 4 * GC.BATTLE.TICK_PER_SECOND;
        this.range = 0.8;
    },
    setUI:function (spellUI){
        this.ui = spellUI;
    },
    update:function (monsters, towers){
        //cc.log("HealSpell update: " + this.curPos.x + " " + this.curPos.y);
        if(this.leftTick > 0){
            for(let i = 0; i < monsters.length; ++i){
                let monster = monsters[i];
                if (Helper.getDistance(this.curPos, monster.curPos) <= this.range && monster.isBuffBySpell(this) == false) {
                    monster.addSpell(this);
                }
            }
            this.leftTick -= 1;
            return true;
        }
        else {
            if(this.ui){
                this.ui.removeFromParent();
            }
            return false;
        }
    },
    removeEffect:function (object){

    },
    doAction:function (object){
        if(object.getEffectTick(this) % 6 == 0){
            object.heal(2);
        }
    },
    setBattleId:function (battleId){
        this.battleId = battleId;
    },
    doEffect:function (monster){

    },
    initAnim:function (){
        let path = this.animPath;
        this.anim = new sp.SkeletonAnimation(path + ".json", path + ".atlas");
        this.anim.setAnchorPoint(cc.p(0.5, 0.5));
        this.anim.setPosition(cc.p(0, 0));
        this.anim.setAnimation(0, this.animArr[0], true);
        this.anim.update(0.0);
        return this.anim;
    }
});
